File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *slow = head , *fast = head ; 
        while (fast && n--) {
            fast = fast->next ; 
        }
        ListNode *prev = nullptr ;
        while (fast) {
            fast = fast->next ; 
            prev = slow ;
            slow = slow->next ;
        }
        if (!prev) {
            return head->next ;
        } else {
            prev->next = slow->next ;
            return head ;
        }
    }
};

File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    var slow = head , fast = head ;
    while (fast && n > 0) {
        fast = fast.next ; 
        n = n - 1 ;
    }
    var prev = null ;
    while (fast) {
        fast = fast.next ; 
        prev = slow ; 
        slow = slow.next ;
    }
    if (!prev) {
        return head.next ;
    } else {
        prev.next = slow.next ;
        return head ;
    }
};